kernel/fs/fat32fs.c

```diff
@@ -0,0 +1,1029 @@
+    struct fat_filesysteminfo fsinfo;
+    struct vfs_node *device;
+    size_t clustersize;
+    size_t clustercount;
+    off_t fatoffset;
+    off_t dataoffset;
+    ino_t currentinode;
+};
+
+struct fat32fs_resource {
+    struct resource;
+    struct fat32fs *fs;
+    struct resource *dir;
+    off_t diroffset;
+    cluster_t cluster;
+};
+
+#define FINALCLUSTER 0xffffff8
+#define IS_FINALCLUSTER(c) (c >= FINALCLUSTER)
+#define DENT_GETCLUSTER(d) ((cluster_t)d->clusterlow | ((cluster_t)d->clusterhigh << 16));
+#define DENT_SETCLUSTER(dent, val) \
+(dent)->clusterlow = val & 0xffff;\
+(dent)->clusterhigh = (val >> 16) & 0xffff;
+
+#define ATTR_READONLY 0x1
+#define ATTR_HIDDEN 0x2
+#define ATTR_SYSTEM 0x4
+#define ATTR_VOLUMEID 0x8
+#define ATTR_DIR 0x10
+#define ATTR_ARCHIVE 0x20
+#define ATTR_DEVICE 0x40
+#define ATTR_LFN (ATTR_VOLUMEID | ATTR_SYSTEM | ATTR_HIDDEN | ATTR_READONLY)
+
+#define TIME_SECONDS(t) ((t & 0b11111) << 2)
+#define TIME_MINUTES(t) (t & 0b11111100000)
+#define TIME_HOUR(t) (t & 0b1111100000000000)
+#define DATE_DAY(d) (t & 0b11111)
+#define DATE_MONTH(d) (t & 0b111100000)
+#define DATE_YEAR(d) (t & 0b1111111000000000)
+
+static cluster_t fat32fs_next(struct fat32fs *fs, cluster_t cluster) {
+    struct resource *res = fs->device->resource;
+    ASSERT(res->read(res, NULL, &cluster, fs->fatoffset + cluster * 4, 4) > 0);
+    return cluster;
+}
+
+static void fat32fs_setnext(struct fat32fs *fs, cluster_t cluster, cluster_t val) {
+    struct resource *res = fs->device->resource;
+    for (int fat = 0; fat < fs->br.fatcount; ++fat) { // change it on all FATs
+        off_t currentfatoffset = fs->fatoffset + fs->br.sectorsperfat * fat * res->stat.st_blksize;
+        ASSERT(res->write(res, NULL, &val, currentfatoffset + cluster * 4, 4) > 0);
+    }
+}
+
+static cluster_t fat32fs_skip(struct fat32fs *fs, cluster_t cluster, size_t count, bool *end) {
+    *end = false;
+    for (size_t i = 0; i < count; ++i) {
+        cluster_t next = fat32fs_next(fs, cluster);
+
+        if (IS_FINALCLUSTER(next)) {
+            *end = true;
+            break;
+        }
+
+        cluster = next;
+    }
+    return cluster;
+}
+
+static size_t fat32fs_getchainsize(struct fat32fs *fs, cluster_t cluster) {
+    size_t count = 0;
+    while (IS_FINALCLUSTER(cluster) == false) {
+        ++count;
+        cluster = fat32fs_next(fs, cluster);
+    }
+    return count;
+}
+
+static off_t fat32fs_clusterdiskoffset(struct fat32fs *fs, cluster_t cluster) {
+    return fs->dataoffset + fs->clustersize * (cluster-2);
+}
+
+static ssize_t fat32fs_rwcluster(struct fat32fs *fs, void *buffer, cluster_t cluster, off_t offset, size_t count, bool write) {
+    return write ?
+            fs->device->resource->write(fs->device->resource, NULL, buffer, fat32fs_clusterdiskoffset(fs, cluster) + offset, count) :
+            fs->device->resource->read(fs->device->resource, NULL, buffer, fat32fs_clusterdiskoffset(fs, cluster) + offset, count);
```

Comment:
this could actually be
```c
(write ? fs->device->resource->write : fs->device->resource->read)(fs->device->resource, NULL, buffer, fat32fs_clusterdiskoffset(fs, cluster) + offset, count);
```

Reply:
That doesn't compile. resource->write and resource->read have different argument types (const void * and void *).

Reply:
oh, fair enough